File: src/platforms/intigriti.py
"""Intigriti Researcher API Connector"""
import os
import httpx
from pathlib import Path
from typing import Optional

BASE_URL = "https://api.intigriti.com/external/researcher/v1"


def _load_token_from_config() -> str:
    """Fallback: read token from config.yaml if env var not set."""
    try:
        import yaml
        config_path = Path(__file__).parent.parent.parent / "config" / "config.yaml"
        if config_path.exists():
            data = yaml.safe_load(config_path.read_text())
            return data.get("platforms", {}).get("intigriti", {}).get("api_key", "")
    except Exception:
        pass
    return ""


class IntigritiClient:
    """Client for the Intigriti Researcher API (v1.0 Beta)."""

    def __init__(self, api_token: Optional[str] = None):
        self.token = api_token or os.environ.get("INTIGRITI_API_TOKEN", "") or _load_token_from_config()
        self.headers = {
            "Authorization": f"Bearer {self.token}",
            "Accept": "application/json",
        }

    async def _get(self, path: str, params: Optional[dict] = None) -> dict:
        async with httpx.AsyncClient(timeout=30) as client:
            resp = await client.get(
                f"{BASE_URL}{path}",
                headers=self.headers,
                params=params or {},
            )
            resp.raise_for_status()
            return resp.json()

    async def get_programs(
        self,
        status_id: Optional[int] = None,
        type_id: Optional[int] = None,
        following: Optional[bool] = None,
        limit: int = 100,
        offset: int = 0,
    ) -> dict:
        """List all programs accessible to the researcher."""
        params = {"limit": limit, "offset": offset}
        if status_id is not None:
            params["statusId"] = status_id
        if type_id is not None:
            params["typeId"] = type_id
        if following is not None:
            params["following"] = str(following).lower()
        return await self._get("/programs", params)

    async def get_program_detail(self, program_id: str) -> dict:
        """Get full program details including scope and rules."""
        return await self._get(f"/programs/{program_id}")

    async def get_activities(
        self,
        created_since: Optional[int] = None,
        following: Optional[bool] = None,
        limit: int = 50,
        offset: int = 0,
    ) -> dict:
        """Get recent program activities (scope/rule changes)."""
        params = {"limit": limit, "offset": offset}
        if created_since is not None:
            params["createdSince"] = created_since
        if following is not None:
            params["following"] = str(following).lower()
        return await self._get("/programs/activities", params)

    async def get_payouts(
        self,
        status_id: Optional[int] = None,
        created_since: Optional[int] = None,
        paid_since: Optional[int] = None,
        limit: int = 100,
        offset: int = 0,
    ) -> dict:
        """Get all payouts (beta)."""
        params = {"limit": limit, "offset": offset}
        if status_id is not None:
            params["statusId"] = status_id
        if created_since is not None:
            params["createdSince"] = created_since
        if paid_since is not None:
            params["paidSince"] = paid_since
        return await self._get("/payouts", params)

    async def get_program_domains(self, program_id: str, version_id: str) -> dict:
        """Get specific version of program domains/scope."""
        return await self._get(f"/programs/{program_id}/domains/{version_id}")

    async def get_program_rules(self, program_id: str, version_id: str) -> dict:
        """Get specific version of rules of engagement."""
        return await self._get(
            f"/programs/{program_id}/rules-of-engagements/{version_id}"
        )

    def extract_compliance(self, program_detail: dict) -> dict:
        """Extract compliance/rules of engagement info from program detail."""
        roe = program_detail.get("rulesOfEngagement") or {}
        # ROE uses version wrapper: {id, createdAt, content: {...}}
        content = roe.get("content") if isinstance(roe, dict) else {}
        if content is None:
            content = {}

        testing = {}
        if isinstance(content, dict):
            testing = content.get("testingRequirements") or {}

        automated = testing.get("automatedTooling")
        # automatedTooling enum: null=unspecified, 1=allowed, 0=not allowed
        if automated is None:
            tooling_status = "unknown"
        elif automated == 1:
            tooling_status = "allowed"
        elif automated == 0:
            tooling_status = "not_allowed"
        else:
            tooling_status = "conditional"

        return {
            "automated_tooling": automated,
            "automated_tooling_status": tooling_status,
            "safe_harbour": content.get("safeHarbour", False) if isinstance(content, dict) else False,
            "user_agent": testing.get("userAgent"),
            "request_header": testing.get("requestHeader"),
            "description": content.get("description", "") if isinstance(content, dict) else "",
            "intigriti_me": testing.get("intigritiMe", False),
        }

    def normalize_program(self, raw: dict) -> dict:
        """Convert Intigriti API program to our DB schema format."""
        domains = []
        if "domains" in raw and "content" in raw.get("domains", {}):
            for d in raw["domains"]["content"]:
                domains.append({
                    "id": d.get("id", ""),
                    "type": d.get("type", {}).get("value", "url"),
                    "asset": d.get("endpoint", ""),
                    "tier": d.get("tier", {}).get("value", ""),
                    "description": d.get("description", ""),
                })

        min_bounty = raw.get("minBounty", {})
        max_bounty = raw.get("maxBounty", {})

        # Extract compliance if rulesOfEngagement is present (detail endpoint)
        compliance = self.extract_compliance(raw) if raw.get("rulesOfEngagement") else None

        return {
            "platform": "INTIGRITI",
            "name": raw.get("name", ""),
            "slug": raw.get("handle", ""),
            "url": raw.get("webLinks", {}).get("detail", ""),
            "intigriti_id": raw.get("id", ""),
            "scope": domains,
            "min_bounty": min_bounty.get("value") if isinstance(min_bounty, dict) else None,
            "max_bounty": max_bounty.get("value") if isinstance(max_bounty, dict) else None,
            "currency": max_bounty.get("currency", "EUR") if isinstance(max_bounty, dict) else "EUR",
            "status": raw.get("status", {}).get("value", ""),
            "type": raw.get("type", {}).get("value", ""),
            "confidentiality": raw.get("confidentialityLevel", {}).get("value", ""),
            "following": raw.get("following", False),
            "industry": raw.get("industry", ""),
            "compliance": compliance,
        }

    async def fetch_all_programs_normalized(self) -> list[dict]:
        """Fetch all programs and normalize them."""
        all_programs = []
        offset = 0
        limit = 100
        while True:
            data = await self.get_programs(limit=limit, offset=offset)
            records = data.get("records", [])
            if not records:
                break
            for r in records:
                all_programs.append(self.normalize_program(r))
            if len(records) < limit:
                break
            offset += limit
        return all_programs

File: dashboard/src/lib/intigriti.ts
/**
 * Intigriti Researcher API client — server-side only.
 * Replaces the Python proxy; Next.js calls Intigriti directly.
 */

import crypto from "crypto";

const DEFAULT_BASE_URL = "https://api.intigriti.com/external/researcher/v1";
const STUB_TOKEN_SENTINEL = "stub-mode";

const BASE_URL = process.env.INTIGRITI_API_URL || DEFAULT_BASE_URL;

function getToken(): string {
  return process.env.INTIGRITI_API_TOKEN || "";
}

/** Mirror of src/reporter/intigriti_submitter.py SEVERITY_MAP. info => null. */
const SEVERITY_MAP: Record<string, string | null> = {
  critical: "Critical",
  high: "High",
  medium: "Medium",
  low: "Low",
  info: null,
};

/** Mirror of src/reporter/intigriti_submitter.py TYPE_MAP. */
const TYPE_MAP: Record<string, string> = {
  idor: "Insecure Direct Object Reference",
  xss: "Cross-Site Scripting",
  sqli: "SQL Injection",
  ssrf: "Server-Side Request Forgery",
  csrf: "Cross-Site Request Forgery",
  access_control: "Broken Access Control",
  info_disclosure: "Information Disclosure",
  differential: "Authorization Flaw",
};

export type FindingPayload = {
  module: string;
  severity: string;
  title: string;
  description: string;
  url?: string | null;
  evidence?: unknown;
  cvssVector?: string | null;
  cweId?: string | null;
};

export type SubmissionResult = {
  submissionId: string | null;
  status: "stubbed" | "submitted" | "deduped" | "rejected_info_severity";
  deduped: boolean;
  stubbed: boolean;
};

export function isStubMode(): boolean {
  const t = getToken();
  return !t || t === STUB_TOKEN_SENTINEL;
}

/** SHA256(programId|module|title|url|cweId) — matches Python `compute_dedup_hash`. */
export function computeDedupHash(finding: FindingPayload, programId: string): string {
  const material = [
    programId,
    finding.module ?? "",
    finding.title ?? "",
    finding.url ?? "",
    finding.cweId ?? "",
  ].join("|");
  return crypto.createHash("sha256").update(material).digest("hex");
}

/**
 * Submit a single confirmed finding to Intigriti.
 * When INTIGRITI_API_TOKEN is missing or 'stub-mode', logs intent and
 * returns a deterministic fake id without any HTTP call.
 */
export async function submitFinding(
  finding: FindingPayload,
  programId: string
): Promise<SubmissionResult> {
  const dedupHash = computeDedupHash(finding, programId);
  const severityLabel = SEVERITY_MAP[(finding.severity || "").toLowerCase()];

  if (severityLabel === null || severityLabel === undefined) {
    return {
      submissionId: null,
      status: "rejected_info_severity",
      deduped: false,
      stubbed: isStubMode(),
    };
  }

  if (isStubMode()) {
    const stubId = `stub-${dedupHash.slice(0, 12)}`;
    console.log(`[STUB] would submit ${finding.title} to ${programId}`);
    return {
      submissionId: stubId,
      status: "stubbed",
      deduped: false,
      stubbed: true,
    };
  }

  const payload: Record<string, unknown> = {
    programId,
    title: finding.title,
    description: finding.description,
    severity: severityLabel,
    type: TYPE_MAP[(finding.module || "").toLowerCase()] || "Other",
  };
  if (finding.url) payload.endpoint = finding.url;
  if (finding.evidence) payload.evidence = finding.evidence;
  if (finding.cvssVector) payload.cvssVector = finding.cvssVector;
  if (finding.cweId) payload.cweId = finding.cweId;

  const res = await fetch(`${BASE_URL}/submissions`, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${getToken()}`,
      "Content-Type": "application/json",
      Accept: "application/json",
    },
    body: JSON.stringify(payload),
    cache: "no-store",
  });

  if (!res.ok) {
    throw new Error(`Intigriti submit failed: ${res.status} ${res.statusText}`);
  }

  const data = (await res.json()) as Record<string, unknown>;
  const submissionId =
    (data.id as string) ||
    (data.submissionId as string) ||
    (data.submission_id as string) ||
    "";
  if (!submissionId) {
    throw new Error("Intigriti response missing submission id");
  }
  return {
    submissionId,
    status: "submitted",
    deduped: false,
    stubbed: false,
  };
}

async function intigritiGet(path: string, params?: Record<string, string>) {
  const url = new URL(`${BASE_URL}${path}`);
  if (params) {
    Object.entries(params).forEach(([k, v]) => {
      if (v) url.searchParams.set(k, v);
    });
  }
  const res = await fetch(url.toString(), {
    headers: {
      Authorization: `Bearer ${getToken()}`,
      Accept: "application/json",
    },
    cache: "no-store",
  });
  if (!res.ok) {
    throw new Error(`Intigriti API error: ${res.status} ${res.statusText}`);
  }
  return res.json();
}

export async function listPrograms(opts: {
  limit?: number;
  offset?: number;
  following?: boolean;
  statusId?: number;
  typeId?: number;
} = {}) {
  const params: Record<string, string> = {
    limit: String(opts.limit ?? 100),
    offset: String(opts.offset ?? 0),
  };
  if (opts.following) params.following = "true";
  if (opts.statusId !== undefined) params.statusId = String(opts.statusId);
  if (opts.typeId !== undefined) params.typeId = String(opts.typeId);
  return intigritiGet("/programs", params);
}

export async function getProgramDetail(programId: string) {
  return intigritiGet(`/programs/${programId}`);
}

export async function getActivities(opts: {
  limit?: number;
  offset?: number;
  following?: boolean;
} = {}) {
  const params: Record<string, string> = {
    limit: String(opts.limit ?? 50),
    offset: String(opts.offset ?? 0),
  };
  if (opts.following) params.following = "true";
  return intigritiGet("/programs/activities", params);
}

export async function getPayouts(opts: {
  limit?: number;
  offset?: number;
} = {}) {
  const params: Record<string, string> = {
    limit: String(opts.limit ?? 100),
    offset: String(opts.offset ?? 0),
  };
  return intigritiGet("/payouts", params);
}

/** Extract compliance/ROE info from program detail response */
export function extractCompliance(programDetail: Record<string, unknown>) {
  const roe = programDetail.rulesOfEngagement as Record<string, unknown> | undefined;
  const content = (roe?.content ?? {}) as Record<string, unknown>;
  const testing = (content.testingRequirements ?? {}) as Record<string, unknown>;

  const automated = testing.automatedTooling as number | boolean | null | undefined;
  let toolingStatus: string;
  if (automated === null || automated === undefined) {
    toolingStatus = "unknown";
  } else if (automated === 1 || automated === true) {
    toolingStatus = "allowed";
  } else if (automated === 0 || automated === false) {
    toolingStatus = "not_allowed";
  } else {
    toolingStatus = "conditional";
  }

  return {
    automatedTooling: automated,
    automatedToolingStatus: toolingStatus,
    safeHarbour: Boolean(content.safeHarbour),
    userAgent: (testing.userAgent as string) || null,
    requestHeader: (testing.requestHeader as string) || null,
    description: (content.description as string) || "",
    intigritiMe: Boolean(testing.intigritiMe),
  };
}

/** Normalize an Intigriti program to our DB format */
export function normalizeProgram(raw: Record<string, unknown>) {
  const domains: unknown[] = [];
  const domainsData = raw.domains as Record<string, unknown> | undefined;
  if (domainsData?.content && Array.isArray(domainsData.content)) {
    for (const d of domainsData.content) {
      const domain = d as Record<string, unknown>;
      const type = domain.type as Record<string, unknown> | undefined;
      const tier = domain.tier as Record<string, unknown> | undefined;
      domains.push({
        id: domain.id || "",
        type: type?.value || "url",
        endpoint: domain.endpoint || "",
        tier: tier?.value || "",
        description: domain.description || "",
      });
    }
  }

  const minBounty = raw.minBounty as Record<string, unknown> | undefined;
  const maxBounty = raw.maxBounty as Record<string, unknown> | undefined;
  const status = raw.status as Record<string, unknown> | undefined;
  const pType = raw.type as Record<string, unknown> | undefined;
  const conf = raw.confidentialityLevel as Record<string, unknown> | undefined;
  const webLinks = raw.webLinks as Record<string, unknown> | undefined;
  const compliance = raw.rulesOfEngagement ? extractCompliance(raw) : null;

  return {
    platform: "INTIGRITI" as const,
    name: (raw.name as string) || "",
    slug: (raw.handle as string) || "",
    url: (webLinks?.detail as string) || "",
    intigritiId: (raw.id as string) || "",
    scope: domains,
    compliance: compliance || {},
    minBounty: minBounty?.value as number | undefined,
    maxBounty: maxBounty?.value as number | undefined,
    currency: (maxBounty?.currency as string) || "EUR",
    industry: (raw.industry as string) || null,
    programType: (pType?.value as string) || null,
    confidentiality: (conf?.value as string) || null,
    active: (status?.value as string)?.toLowerCase() === "open",
  };
}
